### ISSUE TITLE
Add another image proxy to give users alternatives

### ISSUE THREAD

**Issue by galpt** [KotatsuApp/Kotatsu#771]
### Describe your suggested feature

Kind of similar to https://github.com/KotatsuApp/Kotatsu/issues/357.
Some manga websites load images *really* slow.
Image proxies will help boosting image load times from the user's perspective.

Adding wsrv.nl was a good idea, but I think it's better to add more proxies and give users the ability to choose what proxy works better for them.

I would recommend adding [this proxy](https://0ms.dev/mirrors/) too to help speed up loading the images.
This does not reduce image quality, but it caches the images nearest to the users, so it will still help speeding up image load times.

From the user's perspective, useful for users who aren't worried about bandwidth but keep getting slow load times.
From the manga website's perspective, caching proxy will help reducing the server's burden for manga titles that are frequently visited/downloaded.

tl;dr — as the title suggests, this gives another alternative for users to use & maybe compare too.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.

**Comment by galpt** [comment]
I hope this suggestion will be considered or even implemented as the benefits are really that *great*, directly related to the manga images (how fast images get loaded without the need to save them offline).

As this app is about how good it is for reading manga images, I'd argue that feature suggestions affecting the images (i.e. if it's about making images load faster) are more important than other suggestions such as *sorting* or similar issues.

Ofc more discussion about this is great, but I'm sure the above statement is already acceptable as a good reason.

**Comment by galpt** [comment]
To give everyone an idea about how useful is this if it gets implemented, we can compare the manga image URL with & without using the [caching proxy](https://0ms.dev/mirrors/) I mentioned above.

## Scenario: Reading manga images without using image proxy

_Figure 1. The image below shows us that loading all the manga images will be slow for some users, without using any image proxy at all._

![benchmark without 0ms caching proxy](https://github.com/KotatsuApp/Kotatsu/assets/34034590/d6499a12-ba69-4844-9232-3326e6bdef46)

## Comparison: Measuring image load time for a specific manga image URL

_Figure 2. The below image shows us the load time for [specific image URL](https://images2.imgbox.com/9d/09/Dsgt8CzJ_o.jpg), without using any image proxy._

![benchmark without 0ms caching proxy - first try](https://github.com/KotatsuApp/Kotatsu/assets/34034590/604569d4-cb0d-4cb3-8e94-b89043a7cbbf)

_Figure 3. The below image shows us the load time for [specific image URL](https://images2.imgbox.com/9d/09/Dsgt8CzJ_o.jpg), after using [0ms caching proxy](https://0ms.dev/mirrors/images2.imgbox.com/9d/09/Dsgt8CzJ_o.jpg)._

![benchmark with 0ms caching proxy - second try](https://github.com/KotatsuApp/Kotatsu/assets/34034590/ab90851a-288f-4bd1-a8a2-148dcc2e286e)

## How to Use: Image URL structure

_Figure 4. The image URL without using the caching proxy._

![benchmark without 0ms caching proxy - real url](https://github.com/KotatsuApp/Kotatsu/assets/34034590/a8f3b9d7-8bd5-4092-9512-829c3bd9de8c)

_Figure 5. The image URL after using the caching proxy._

![0ms caching proxy - how to use](https://github.com/KotatsuApp/Kotatsu/assets/34034590/6d72ef62-6d32-4288-a094-4696951cc191)

**Comment by galpt** [comment]
For caching so many images (I mean, _millions_ of images), it is recommended to use the [x.0ms.dev](https://x.0ms.dev/) endpoint, as described on the web page (https://0ms.dev/mirrors/).

The reasons are:
1. Server's network is fast (at least 4 Gbit/s guaranteed). Really good peering (almost as good as Cloudflare).
2. The caching program is custom-made, designed to be low latency (in-memory caching, and other optimizations).
3. For that endpoint, it doesn't rely on Cloudflare. This is to make sure the caching proxy doesn't get throttled by Cloudflare, in case Kotatsu users will make 100 TB of image traffic or even more.

- Reason (1) is good to make sure network connection between users' ISPs and the proxy is good (a.k.a good peering).
- Reason (2) will help a lot for use cases like caching a lot of images (i.e. for manga reader/downloader apps).
- Reason (3) is needed because according to some sources (reddit, etc.) Cloudflare won't like it if we use their service to cache lots of images. That's why the caching proxy should not rely on them.

<details>
  <summary>fyi</summary>
  Btw, I'm one of the devs that designed the 0ms caching proxy, so that's why I know some details such as point 1-3.
</details>

tl;dr — completely okay when using the [x.0ms.dev](https://x.0ms.dev/) endpoint. The endpoint is specifically designed for lots of images.

**Comment by Koitharu** [comment]
Thank you for so detailed explanation, I'll implement it soon

**Comment by galpt** [comment]
For Kotatsu app, you can use either `0ms.dev/mirrors/` or `x.0ms.dev/q70/`.
Both endpoints are handled by the same servers, but it's only that `x.0ms.dev/q70/` supports getting images from `http://` URLs, while `0ms.dev/mirrors/` will automatically use `https://` for all URLs.

**Comment by galpt** [comment]
Hoping it'll be implemented very soon in the next major releases.

This will make https://github.com/KotatsuApp/Kotatsu/issues/640 kind of obsolete.
It does some optimizations to the cached files without compressing the files, so technically images will load faster without reducing the quality.
Very useful if the user's phone doesn't have big RAM to keep the cache in memory or bloating the storage but the user still wants to make images load faster. Home users with fast WiFi will love it.

**Comment by Koitharu** [comment]
I added this proxy however looks like it slow down requests significantly comparing to wsrv.nl and direct requests...

**Comment by galpt** [comment]
@Koitharu 
Can you give me a real example (i.e. the specific image links you're testing)?

Technically, it will be slower _only the first time_ because the proxy need to download every single image, save them to the storage servers, and cache them globally.
But anything that's already in the storage will be served faster and won't even reach the original image link (unproxied link) for the next 7 days.

The difference between wsrv.nl is that they don't guarantee your images will be cached, they may have a limited cache for other images.  The 0ms.dev image proxy is a guaranteed cache, repetitive requests of the same image link(s) will be faster compared to accessing the image via direct link

**Comment by galpt** [comment]
I've tested the latest version (v7.1) and the 0ms mirror is working as it should.

All images are cached (downloaded & stored to storage) with the average size ranging from 500 KB to 1 MB per-image (could be more, depends on the image height+width & quality).

Just like how any caching logic works, the very 1st time we are accessing it, we're going to feel some delay, that means the content was not in the cache and we are telling the logic to cache it.

The 2nd time and the n-time we're accessing it, it will be faster, could be almost instant if our connection to Cloudflare is really good.


As the number of users who are using the mirror increases, the number of cached images will also increase, so the same manga images will definitely load faster.. This is a good thing and it will get better over time.
We're feeling the little delays because we're the first users who're testing it :)

**Comment by galpt** [comment]
I'm closing this issue as the feature's already implemented. Thank you!

**Comment by Ero-gamer** [comment]
> I'm closing this issue as the feature's already implemented. Thank you!

Your proxy server is amazing and very fast. Can I use your proxy as a DNS resolver like cloudflare DNS 1.1.1.1? If yes, what is 0ms DNS resolver address ?

**Comment by galpt** [comment]
Hi @Ero-gamer , thanks for testing the proxy.
The DNS resolver is available on [homepage](https://0ms.dev/), feel free to use it

**Comment by Berserk0001** [comment]
> Hoping it'll be implemented very soon in the next major releases.
> 
> This will make [#640](https://github.com/KotatsuApp/Kotatsu/issues/640) kind of obsolete. It does some optimizations to the cached files without compressing the files, so technically images will load faster without reducing the quality. Very useful if the user's phone doesn't have big RAM to keep the cache in memory or bloating the storage but the user still wants to make images load faster. Home users with fast WiFi will love it.

I don't think that bandwidth-hero-proxy is obsolete 'cause I have been using for 1 year on TachiyomiSY and it is very fast for me. And i don't pay for the servers since vercel's free tier is enough.

**Comment by galpt** [comment]
@pawankushum what site? 0ms migrated to v.recipes, but it's already updated in kotatsu since May
